src/foam/core/AbstractAgency.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'AbstractAgency',
+  abstract: true,
+  implements: [ 'foam.core.Agency' ],
+
+  javaImports: [
+    'foam.nanos.logger.Loggers',
+    'java.util.Timer',
+    'java.util.TimerTask',
+    'java.util.concurrent.ConcurrentHashMap'
+  ],
+
+  javaCode: `
+    protected static final Timer TIMER = new Timer();
+    protected static final ConcurrentHashMap<String, TimerTask> TASK_QUEUE = new ConcurrentHashMap<>();
+
+    public void schedule(X x, ContextAgent agent, String key, long delay) {
+      if ( delay <= 0 ) {
+        submit(x, agent, key);
+        return;
+      }
+
+      var task = TASK_QUEUE.get(key);
+
+      // Cancel the task if it's not yet executed and being re-scheduled
+      if ( task != null
+        && task.scheduledExecutionTime() - System.currentTimeMillis() > 0
+      ) {
+        task.cancel();
+        Loggers.logger(x, this).debug("schedule", "cancel", key);
+      }
+
+      // Schedule executing the agent
+      task = new TimerTask() {
+        @Override
+        public void run() {
+          try { agent.execute(x); }
+          catch ( Exception e ) {
+            Loggers.logger(x, this).error("schedule", "failed", key);
```

Comment:
put the exception in the log 
```suggestion
            Loggers.logger(x, this).error("schedule", "failed", key, e);
```

Reply:
Question: will this change the behaviour of async rules (or rules in general) in terms of retrying a failed task since the exception is no longer thrown to the calling code?

Reply:
agency.schedule() will not retry same for agency.submit() but the rule engine (rule.asyncApply) does handle retry for async rules. 

We would have to re-enter rule.asyncApply in the RuleActionMergeDecorator to support retry https://github.com/kgrgreer/foam3/pull/1027/commits/7dafccd34cfb8820150867218fc4062dbbc7a4a2, @nanoscott, @jlhughes @kgrgreer, thoughts?